### PR TITLE
Member only app v0.0.1

### PR DESCRIPTION
**Pull request by YoseptF**
> ON this project we made an application where you can post things if you are part of the club (_verified with your login information_), then members can see both, what was posted and by whom, but if you're not part of the clubhouse, you can only wonder with the comments.

You can either clone it or use the live version [here](https://intense-sands-50703.herokuapp.com/)

These are the principal characteristics:

- [x] User:
    - Can post comments
    - Has many posts
- [x] Posts:
    - Belong to a user
- [x] Logged in users can see both messages and creator
- [x] Non-Logged in users can see both messages

---

As always thanks for your time, and we wait for your comments, have a nice day.

### CODE REVIEW

File: app/controllers/main_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class MainController < ApplicationController
+  def index
+    @posts = Post.paginate(page: params[:page], per_page: 10)
+  end
```
Comment (maelfosso) [inline app/controllers/main_controller.rb]: Move this function in `PostsController` and delete that controller

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class PostsController < ApplicationController
+  def new
+    @post = Post.new
+  end
+
+  def create
+    user = session[:user_id]
```
Comment (maelfosso) [inline app/controllers/posts_controller.rb:13]: Forget about `session` usage

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello coders,
I have nothing to say but congratulates you. The result is awesome :+1: 
However, allow me, please, to mention some requirements that need to be met

1. It has been said [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/bcrypt_requirement.png), The application uses the bcrypt gem for authentication. Sorry but you didn't use it

2. Please, because of [that](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_method_requirement.png), please forget about using `session` and implement authentication using **cookies**

3. As [Indicated here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png), you should uses the #current_user method to retrive user  and you should uses the #current_user=(user) method to set user. But one is missing and the other one is not well implemented

4. From [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png), It creates the application routes using `resources :posts, only: [:new, :create, :index]`, but there is no routes `Posts#index`

5. I'm not connected but I can access to `Post#new` to create post
Screenshot from 2020-03-02 19-40-53
While it was said [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) that it must not be possible

5. Read the comment below

Kindly make the suggested changes and submit a new code review request

Mael FOSSO

**Comment by YoseptF** [comment]
Hi, @maelfosso,
We have made the changes required, now listed here:
- [x] Use the `Bcrypt gem` to authenticate
image
- [x] The current method now returns the correct current_user
image
- [x] Only logged in users can post now
image
image
- [x] We moved, even tho we don't think that was neccesary, the query from `main_controller` to `posts_controller` and deleted the main_controller
image
image

As always, thank you for your time, and we await your comments.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello, thank you for the improvement done!
However I think that you missunderstand a requirement. 

> The remember_me token the project describes consists of storing a hash in a cookie and using it to determine in a user is logged in. Do not confuse it with the Remember me [ ] checkbox that many websites show. That is a more complex solution, which is not required here.

That is why you should focus only on **`cookies`** for authentication instead of `session`. 
For a better understanding, when the user sign in, you have to set its remember token. So instead of 
```ruby
def log_in(user)
    session[:user_id] = user.id
  end
```
You should have something like
```ruby
# 1. Generate the digest remember token
# 2. You update it in the user who correctly sign in
# 3. You set the cookies with the remember token
# 3. You set the current user
```
With that vision of the work, you will simplify your `def current_user` method to 1 or 2 lines.

Also, sorry, I have forgot to mention it in the last review, but [as indicated here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_create_requirement.png) you forgot the `before_create` callback. Please, make a difference between `before_create` and `before_save`

Kindly make the suggested changes and submit a new code review request.
Thanks

Mael FOSSO

**Comment by YoseptF** [comment]
Hi @maelfosso,
We want to clarify something, in the requirements it says
```
...That is a more complex solution, which is not required here.
```
but it doesn't seem to forbid it, and we are doing both, so we don't really understand what the problem is here
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/helpers/sessions_helper.rb#L17-L30

---

As for the other problem, you suggest us to follow something like this:

```
# 1. Generate the digest remember token
# 2. You update it in the user who correctly signs in
# 3. You set the cookies with the remember token
# 3. You set the current user
``` 

And we do:
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/helpers/sessions_helper.rb#L5-L11
- [x] Generate the digest remember token
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/helpers/sessions_helper.rb#L7
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/models/user.rb#L22-L27
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/models/user.rb#L12-L15
- [x] You update it in the user who correctly signed in
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/models/user.rb#L25
- [x] You set the cookies with the remember token
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/helpers/sessions_helper.rb#L8-L9
- [x] You set the current user
 this is the only step that's different because we didn't force the first time to remember in the browser since next time it asks for the current user it's gonna set it automatically anyway
https://github.com/YoseptF/Members-Only/blob/b85cf95e3514ae2b73b597dc1b04a337fa977912/app/helpers/sessions_helper.rb#L24
but in the next update, we also made it remember the first time.

**Comment by YoseptF** [comment]
We made the following changes:
- [x] The user now get remembered the first time he signs up
- [x] the email now get an extra validation
- [x] We explained better how our authentication system works

We also want to remember this line from the Odin project itself

image

**Review by Ispirett — APPROVED** [review]
Status Complete :+1: 
Great work folks with all the changes made. you can move on now :100:

**Pull request merged**